From 0766551dd0298b5216699abf3f805c59b92c7840 Mon Sep 17 00:00:00 2001
From: Andrew Hewus Fresh <andrew@afresh1.com>
Date: Fri, 15 Jun 2018 15:47:49 -0700
Subject: [PATCH] Extract files with O_EXCL

This will cause the open to die if the file exists, avoiding a race
condition and security risk.

Relates to commit ae65651eab053fc6dc4590dbb863a268215c1fc5
and https://rt.cpan.org/Ticket/Display.html?id=125523
---
 lib/Archive/Tar.pm | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/lib/Archive/Tar.pm b/lib/Archive/Tar.pm
index 5950b3e..a3aa577 100644
--- a/lib/Archive/Tar.pm
+++ b/lib/Archive/Tar.pm
@@ -861,7 +861,7 @@ sub _extract_file {
     }
     if( length $entry->type && $entry->is_file ) {
         my $fh = IO::File->new;
-        $fh->open( $full, '>' ) or (
+        $fh->open( $full, O_WRONLY|O_CREAT|O_EXCL  ) or (
             $self->_error( qq[Could not open file '$full': $!] ),
             return
         );
-- 
2.16.4

